### DISCUSSION TITLE
Unbounded objective function

### DISCUSSION THREAD

**Discussion by Redkaido** [yalmip/YALMIP#1377] [General discussion] [answered]
I run the code, it shows that YALMIP problem 2 'Unbounded objective function (GUROBI)'
Meanwhile, I have constrained the lb and ub of parameters,  but I check lb and ub of model are inf.
Please help me to solve the problems. Thank you.
![Screenshot 2024-03-03 175616](https://github.com/yalmip/YALMIP/assets/54225043/4b5e54e1-6b3a-418c-8cd1-c86b074cc756)

The matlab code is as follows.

      base=4;
      use=3;
      SNR0 = 100;
      B=499.2e6;
      c = physconst('LightSpeed'); 
      %parameter
      baseloc=intvar(base, 3, 'full');
      useloc=intvar(use, 3, 'full');
      %intermedia
      H=sdpvar(use*base,7, 'full');
      C=sdpvar(7,7);

     cons=[0<=baseloc(1,1)<=500];
     cons=[cons 0<=baseloc(1,2)<=300];
     cons=[cons 0<=baseloc(1,3)<=1500];
     cons=[cons 0<=baseloc(2,1)<=500];
     cons=[cons 0<=baseloc(2,2)<=300];
     cons=[cons 0<=baseloc(2,3)<=1500];
     cons=[cons 0<=baseloc(3,1)<=500];
     cons=[cons 0<=baseloc(3,2)<=300];
     cons=[cons 0<=baseloc(3,3)<=1500];
     cons=[cons 0<=baseloc(4,1)<=500];
     cons=[cons 0<=baseloc(4,2)<=300];
     cons=[cons 0<=baseloc(4,3)<=1500];
 
     cons=[cons 0<=useloc(1,1)<=4600];
     cons=[cons 400<=useloc(1,2)<=520];
     cons=[cons 200<=useloc(1,3)<=1600];
     cons=[cons 0<=useloc(2,1)<=4600];
     cons=[cons 400<=useloc(2,2)<=520];
     cons=[cons 200<=useloc(2,3)<=1600];
     cons=[cons 0<=useloc(3,1)<=4600];
     cons=[cons 400<=useloc(3,2)<=520];
     cons=[cons 200<=useloc(3,3)<=1600];

      for i=1:base
          for j=i+1:base
              cons=[cons sum((baseloc(i,:)-baseloc(j,:)).^2)>=1];
          end
      end
      for i=1:use
          for j=i+1:use
              cons=[cons sum((useloc(i,:)-useloc(j,:)).^2)>=1];
          end
      end
      
      for i=1:3
          cons=[cons sum((baseloc(2:end,i)-kron(ones(base-1,1),baseloc(1,i))).^2)>=1];
          cons=[cons sum((useloc(2:end,i)-kron(ones(use-1,1),useloc(1,i))).^2)>=1];
      end
      centerLoc0=[2300,1300,800];

    q=[0,0,0,1];
    qin1 = [2 * q(1) 2 * q(2) 2 * q(3);
            2 * q(2) -2 * q(1) -2 * q(4);
            2 * q(3) 2 * q(4) -2 * q(1)];
    qin2 = [-2 * q(2) 2 * q(1) 2 * q(4);
            2 * q(1) 2 * q(2) 2 * q(3);
            -2 * q(4) 2 * q(3) -2 * q(2)];
    qin3 = [-2 * q(3) -2 * q(4) 2 * q(1);
            2 * q(4) -2 * q(3) 2 * q(2);
            2 * q(1) 2 * q(2) 2 * q(3)];
    qinw = [2 * q(4) -2 * q(3) 2 * q(2);
            2 * q(3) 2 * q(4) -2 * q(1);
            -2 * q(2) 2 * q(1) 2 * q(4)];
    for idx=1:use
        for jdx=1:base
         %H
    cons=[cons,H(base*idx+jdx-base,:)==[(15.55 -20*log10(sum((baseloc(jdx,:) - useloc(idx,:)).^2, 2)/1000/1000)) *(useloc(idx,:) - baseloc(jdx,:)) * (qin1 * (useloc(idx,:)-centerLoc0)'/1000) / ((sum((baseloc(jdx,:) - useloc(idx,:)).^2, 2))^(0.5)/1000),...
        (15.55 -20*log10(sum((baseloc(jdx,:) - useloc(idx,:)).^2, 2)/1000/1000)) *(useloc(idx,:) - baseloc(jdx,:)) * (qin2 * (useloc(idx,:)-centerLoc0)'/1000) / ((sum((baseloc(jdx,:) - useloc(idx,:)).^2, 2))^(0.5)/1000),...
        (15.55 -20*log10(sum((baseloc(jdx,:) - useloc(idx,:)).^2, 2)/1000/1000)) *(useloc(idx,:) - baseloc(jdx,:)) * (qin3 * (useloc(idx,:)-centerLoc0)'/1000) / ((sum((baseloc(jdx,:) - useloc(idx,:)).^2, 2))^(0.5)/1000),...
        (15.55 -20*log10(sum((baseloc(jdx,:) - useloc(idx,:)).^2, 2)/1000/1000)) *(useloc(idx,:) - baseloc(jdx,:)) * (qinw * (useloc(idx,:)-centerLoc0)'/1000) / ((sum((baseloc(jdx,:) - useloc(idx,:)).^2, 2))^(0.5)/1000),...
        (15.55 -20*log10(sum((baseloc(jdx,:) - useloc(idx,:)).^2, 2)/1000/1000)) *(useloc(idx,1) - baseloc(jdx,1)) / ((sum((baseloc(jdx,:) - useloc(idx,:)).^2, 2))^(0.5)/1000),...
        (15.55 -20*log10(sum((baseloc(jdx,:) - useloc(idx,:)).^2, 2)/1000/1000)) *(useloc(idx,2) - baseloc(jdx,2)) / ((sum((baseloc(jdx,:) - useloc(idx,:)).^2, 2))^(0.5)/1000),...
        (15.55 -20*log10(sum((baseloc(jdx,:) - useloc(idx,:)).^2, 2)/1000/1000)) *(useloc(idx,3) - baseloc(jdx,3)) / ((sum((baseloc(jdx,:) - useloc(idx,:)).^2, 2))^(0.5)/1000)]];    
        end
    end
    cons=[cons, C*(H'*H)==eye(7)];
    Obj=trace(C); %trace((H'H)^-1)
    
    ops = sdpsettings('solver','gurobi+','verbose',2,'gurobi.NonConvex',2,'gurobi.ResultFile','MyFile.lp','debug',1);
    ops.gurobi.TuneTimeLimit = 0;
    result = optimize(cons, Obj, ops);
    if result.problem==0
          value(x)
          value(Obj)
    else
         [model,recoverymodel] = export(cons, Obj, ops);
    end

**Comment by johanlofberg** [comment, marked as answer]
This problem cannot be solved using gurobi. You are forcing yalmip to send the model to gurobi, despite gurobi having no clue what do with it (by using 'gurobi+')

**Reply by Redkaido** [reply]
Dear prof. Johan,
Would you recommend a solver to do it? 
I tried to use fmincon. it only can get a feasible solution in matlab, and it can't have a solution by yalmip+matlab.

**Reply by johanlofberg** [reply]
In YALMIP none, needs better formulation

**Reply by Redkaido** [reply]
Dear prof. Johan,

Excuse me. 
If I changed the OBJ=norm(H). Can I use YALMIP to get the optimum? 

Thank you.
Jun,
NTU

**Reply by johanlofberg** [reply]
Depends on what context you are talking about. With gurobi, norm(H) is reasonable as it is directly socp representable. For a nonlinear solver, it is a bad model, since nonlinear solvers don't like socp models, and should thus simply use the quadratic objective H'*H

**Reply by Redkaido** [reply]
Thank you very much. 
The objective has better directly socp. The Obj=det(H'*H) is also unavailable?

**Reply by johanlofberg** [reply]
there is no way to get that into a nice form https://yalmip.github.io/tutorial/maxdetprogramming/
